apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/peer/FileDownloader.java

```diff
@@ -17,48 +14,90 @@
 import com.apollocurrency.aplwallet.apl.core.peer.statcheck.PeersList;
 import com.apollocurrency.aplwallet.apl.util.ChunkedFileOps;
 import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Base64;
 import java.util.Collection;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
 import javax.enterprise.inject.Vetoed;
 import javax.inject.Inject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /**
  * This class performs complete file downloading from peers
+ *
  * @author alukin@gmail.com
  */
 public class FileDownloader {
+
     @Vetoed
-    public class Status{
-        double completed;
-        int chunksTotal;
-        int chunksReady;
-        List<String> peers; 
+    public class Status {
+        double completed = 0.0;
+        int chunksTotal = 1; //init to 1 to avoid zero division
+        int chunksReady = 0;
+        List<String> peers = new ArrayList<>();
+        FileDownloadDecision decision = FileDownloadDecision.NotReady;
+        boolean isComplete(){
+            return chunksReady==chunksTotal;
+        }
     }
-    
+
+    public static final int DOWNLOAD_THREADS = 6;
     private String fileID;
-    private FileDownloadInfo downloadInfo;    
+    private FileDownloadInfo downloadInfo;
     private List<HasHashSum> goodPeers;
     private List<HasHashSum> badPeers;
-    private Status status;
-    
-    @Inject
+    private final Status status = new Status();
+    private static final Logger log = LoggerFactory.getLogger(FileDownloader.class);
+
     DownloadableFilesManager manager;
-    
-    public FileDownloader() {      
+
+    ExecutorService executor;
+    List<Future<Boolean>> runningDownloaders = new ArrayList<>();
+
+    @Inject
+    public FileDownloader(DownloadableFilesManager manager) {
+        this.manager = manager;
+        this.executor = Executors.newFixedThreadPool(DOWNLOAD_THREADS);
```

Comment:
Add @PreDestroy method for executor shutdown

Reply:
Thanks, it is imortant. fixed.